src/xrpld/app/misc/WasmHostFuncImpl.cpp

```diff
@@ -36,198 +36,409 @@ WasmHostFunctionsImpl::getParentLedgerTime()
+}
+
+int32_t
+WasmHostFunctionsImpl::getCurrentLedgerObjArrayLen(SField const& fname)
+{
+    if (fname.fieldType != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+
+    auto const sle = ctx.view().read(leKey);
+    if (!sle)
+        return HF_ERR_LEDGER_OBJ_NOT_FOUND;
 
-    Slice const s = (*jv)[sfURI];
-    return Bytes{s.begin(), s.end()};
+    auto const* field = sle->peekAtPField(fname);
+    if (!field)
+        return HF_ERR_FIELD_NOT_FOUND;
+
+    if (field->getSType() != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+    int32_t const sz = static_cast<STArray const*>(field)->size();
+
+    return sz;
+}
+
+int32_t
+WasmHostFunctionsImpl::getLedgerObjArrayLen(
+    int32_t cacheIdx,
+    SField const& fname)
+{
+    if (fname.fieldType != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+
+    if (cacheIdx < 0 || cacheIdx >= MAX_CACHE)
+        return HF_ERR_SLOT_OUT_RANGE;
+
+    if (!cache[cacheIdx])
+        return HF_ERR_INVALID_SLOT;
+
+    auto const* field = cache[cacheIdx]->peekAtPField(fname);
+    if (!field)
+        return HF_ERR_FIELD_NOT_FOUND;
+
+    if (field->getSType() != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+    int32_t const sz = static_cast<STArray const*>(field)->size();
+
+    return sz;
+}
+
+int32_t
+WasmHostFunctionsImpl::getTxNestedArrayLen(Slice const& locator)
+{
+    auto const r = locateField(&ctx.tx, locator);
+    if (!r.has_value())
+        return r.error();
+    auto const* field = r.value();
+
+    if (field->getSType() != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+    int32_t const sz = static_cast<STArray const*>(field)->size();
+
+    return sz;
+}
+
+int32_t
+WasmHostFunctionsImpl::getCurrentLedgerObjNestedArrayLen(Slice const& locator)
+{
+    auto const sle = ctx.view().read(leKey);
+    if (!sle)
+        return HF_ERR_LEDGER_OBJ_NOT_FOUND;
+    auto const r = locateField(sle.get(), locator);
+    if (!r.has_value())
+        return r.error();
+    auto const* field = r.value();
+
+    if (field->getSType() != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+    int32_t const sz = static_cast<STArray const*>(field)->size();
+
+    return sz;
+}
+
+int32_t
+WasmHostFunctionsImpl::getLedgerObjNestedArrayLen(
+    int32_t cacheIdx,
+    Slice const& locator)
+{
+    if (cacheIdx < 0 || cacheIdx >= MAX_CACHE)
+        return HF_ERR_SLOT_OUT_RANGE;
+
+    if (!cache[cacheIdx])
+        return HF_ERR_INVALID_SLOT;
+
+    auto const r = locateField(cache[cacheIdx].get(), locator);
+    if (!r.has_value())
+        return r.error();
+
+    auto const* field = r.value();
+
+    if (field->getSType() != STI_ARRAY)
+        return HF_ERR_NO_ARRAY;
+    int32_t const sz = static_cast<STArray const*>(field)->size();
+
+    return sz;
 }
 
-bool
+int32_t
 WasmHostFunctionsImpl::updateData(Bytes const& data)
 {
-    if (!ctx.view().exists(leKey))
-        return false;
     auto sle = ctx.view().peek(leKey);
+    if (!sle)
+        return HF_ERR_LEDGER_OBJ_NOT_FOUND;
     sle->setFieldVL(sfData, data);
```

Comment:
should return `HF_ERR_DATA_FIELD_TOO_LARGE` if data is too large, right?

Reply:
It returns HF_ERR_BUFFER_TOO_SMALL if provided buffer is not large enough.
Maybe  BUFFER_TOO_SMALL and DATA_FIELD_TOO_LARGE are conflicting. BUFFER_TOO_SMALL come first so lets use it?

Reply:
they are for different errors. we have a cap of how much space the user data can take, 4KB. the user cannot store unlimited data even they pay. This DATA_FIELD_TOO_LARGE error is for cases that user tried to store more than 4k.